### ISSUE TITLE
Json/Ion stream parsing

### ISSUE THREAD

**Issue by 9il** [ikod/dlang-requests#133]
Hi @ikod 
Would you accept [Mir Ion](https://github.com/libmir/mir-ion) dependency that will allow to buffered parsing JSON/Ion responses (without allocating the an array for the whole response)? Mir Ion DOM uses Amazon's Ion format for JSON parsing and its DOM is much more smaller than an input data because of symbol tables.

**Comment by ikod** [comment]
Hello @9il 

Will try

**Comment by ikod** [comment]
Hello, @9il 

Do you mean - add Mir Ion dependency to requests library and internally support serde with it based on content-type?
Or something else?

**Comment by 9il** [comment]
I mean fast direct reading from the `Buffer` .

Mir Ion don't have complete buffered API yet. Only buffered `deserializeJsonFile`, but I will add it in case it could be used in the requests.

serde support of content-type can be added on top of it thought

**Comment by ikod** [comment]
Hello @9il 

> I mean fast direct reading from the `Buffer` .
> 
> Mir Ion don't have complete buffered API yet. Only buffered `deserializeJsonFile`, but I will add it in case it could be used in the requests.
> 
> serde support of content-type can be added on top of it thought

I see. So, answering initial question - yes, I will accept dependency from Mir Ion. Streamed parsing is nice and cool feature.

If I can make something right now - I'm ready and need more details

**Comment by 9il** [comment]
@ikod looks like I have misunderstood Buffer implementation. I thought it is kind of buffer stream, but naturally it is an array of already allocated memory chunks.

I am afraid it may require too much efforts on dlang-requests side. The idea is to make an implementation is more lazy and follow the buffer stream concept where the element of stream is a memory chunks. Stack memory or fixed size buffers + delegates/lambdas can be used in this case quite well: streams fills memory into one or more buffer  for async case.

Example:
1. JSON parser requests the next 4 KiB of input text.
2. The underlaying buffer stream fills 4 KiB of JSON parser provided buffer. Or, alternatively, provides it own fixed size, which JSON parser can copy data from.

Also such buffer streams can be used in pipeline with decompressions.

If one send a 112 MiB JSON file like in https://github.com/kostya/benchmarks#json, than total memory required to build its full DOM would be simply 16 MiB. Currently it will require 112 (request body) + 16 MiB (DOM).

**Comment by ikod** [comment]
> I thought it is kind of buffer stream, but naturally it is an array of already allocated memory chunks.

TL/DR: This is DataPipe which built on top of Buffer. Buffer is just a 'sliding window' into received data. It attach new chunks to the tail and free processed parts from the head.

This is how stream processing works in requests in more details:

- "receiveResponse" take raw data from network  in byte array: https://github.com/ikod/dlang-requests/blob/2060d44b968a58170250dcab35ba3508284db1da/source/requests/http.d#L911 receiveResponce provide ubyte[_bufferSize] for data, but we cannot expect anything specific about the length of the received data.

-  data passed to bodyDecoder which is actually chain of decoders: https://github.com/ikod/dlang-requests/blob/2060d44b968a58170250dcab35ba3508284db1da/source/requests/http.d#L933 each decoder append data to internal Buffer, process it if possible (unchunk or uncompress when buffer keeps complete data in some meaning) - and pass processed data down to the next decoder,  then free processed memory chunks: https://github.com/ikod/dlang-requests/blob/2060d44b968a58170250dcab35ba3508284db1da/source/requests/streams.d#L161

- decoders finally produce ubyte[] will be attached to responseBody https://github.com/ikod/dlang-requests/blob/2060d44b968a58170250dcab35ba3508284db1da/source/requests/http.d#L942

- in "streamed" mode we permanently put and get data from decoders:
```d
    __bodyDecoder.putNoCopy(b[0..read]);
    auto res = __bodyDecoder.getNoCopy(); 
    if ( res.length == 0 ) {   // there were nothing to produce (beginning of the chunk or no decompressed data)            
        continue;
    }
```

So this is something like chain of data converters. It is composable, pluggable and allow avoid data copy when possible.

> * JSON parser requests the next 4 KiB of input text.
> * The underlaying buffer stream fills 4 KiB of JSON parser provided buffer. Or, alternatively, provides it own fixed size, which JSON parser can copy data from.

Looks like this is how content-encoding and transfer encoding works in `requests` but they consume and produce memory chunks in 'asynchronous' manner, not some data structures.

I think it is possible to decode json stream, but do we need some kind of callbacks to user code, like for xml? How you plan to process unparsed json?

**Comment by ikod** [comment]
Hi @9il 

BTW there are more ways to use streaming parser:

It is possible to use

```d
import std.stdio;
import requests;

void main()
{
    auto rq = Request();
    rq.useStreaming = true;
    rq.verbosity = 2;
    auto rs = rq.get("http://httpbin.org/image/jpeg");
    auto stream = rs.receiveAsRange();
    while(!stream.empty) {
        writefln("Received %d bytes, total received %d from document legth %d", stream.front.length, rs.contentReceived, rs.contentLength);
        stream.popFront;
    }
}
```
where you have access to each response chunk and can parse it chunk by chunk.

I will check if Interceprors(kind of middleware) can be used.

**Comment by 9il** [comment]
> Looks like this is how content-encoding and transfer encoding works in `requests` but they consume and produce memory chunks in 'asynchronous' manner, not some data structures.

I can't get it. All computations look eager, not async. Should buffer consume all input before it can be used for reading? If no, how it is synchronised, I didn't see related mutex usage around to make it async.

> BTW there are more ways to use streaming parser:

Yes, we need something like that.

**Comment by ikod** [comment]
Hi, @9il 

> > Looks like this is how content-encoding and transfer encoding works in `requests` but they consume and produce memory chunks in 'asynchronous' manner, not some data structures.
> 
> I can't get it. All computations look eager, not async. Should buffer consume all input before it can be used for reading? If no, how it is synchronised, I didn't see related mutex usage around to make it async.

They are 'asynchronous' in such sense that reads from network are not synchronized with "record" boundaries in content. For example two lines "01234567\n89abcdef\n" when encoded in HTTP "transfer-encoding: chunked" can be transferred over network like this (two 9-bytes chunks):

```
9
01234567\n
9
89abcdef\n
0
```
and read(socket) can return incomplete part like
```
9
01234567\n
9
89
```
so you can process first line "01234567\n", throw it away, and probably free its memory. For **efficient** handling of remaining part (**<I>zero memory copy, no new allocations</I>**) I save unprocessed parts in Buffer.

I also have even more efficient @safe and @nogc package nbuff, which solve the same problem.

> 
> > BTW there are more ways to use streaming parser:
> 
> Yes, we need something like that.

Great!

**Comment by ikod** [comment]
hi @9il 

Should I close this issue, or there is something to do?

**Comment by 9il** [comment]
Lets close it for now

**Comment by ikod** [comment]
Ok, thanks, @9il
